### ISSUE TITLE
jq error in Certbot snap

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#8280]
On a clean Debian 10 VPS from DigitalOcean I ran:
```
apt update
apt install snapd -y
snap install --classic certbot
/snap/bin/certbot --version
```
The last command fails with the output:
```
jq: error (at <stdin>:0): Cannot iterate over null (null)
```
and the exit status 5.

I'm not sure how widespread the problem is here or what the root cause is, but I wasn't able to reproduce it on my Ubuntu 20.04 machine.

cc @adferrand, @alexzorin

EDIT: After the discussion below, the changes we decided to make here are:

1. ~~Help users avoid installing the Certbot snap with an old version of snapd. Probably the best way to do this it update our instructions on https://certbot.eff.org to have people manually update it, but I personally think it's worth a little time to see if there's an easy way for us to automate this (outside of trying to run `snap install core` ourselves in our snap which I don't think we should do).~~ Done in https://github.com/certbot/website/pull/650.
2. ~~Improve the error message for people who somehow get themselves into this situation anyway. I still think it should be fatal but we can certainly make the description of the problem and how to fix it more clear. As part of this, I think we may want to consider rewriting `certbot.wrapper` in Python as suggested by Alex at https://github.com/certbot/certbot/issues/8280#issuecomment-690768231, but I personally think we can leave whether or not this is done up to whoever implements this.~~ Done in https://github.com/certbot/certbot/pull/8313.
3. ~~Report a bug against https://github.com/canonical-web-and-design/snapcraft.io asking them to fix the Debian instructions to avoid this problem.~~ Done in https://github.com/canonical-web-and-design/snapcraft.io/pull/3074.

**Comment by alexzorin** [comment]
I can repro too.

    root@history-episode:/snap/certbot/current# curl -s --unix-socket /run/snapd.socket "http://localhost/v2/connections?snap=certbot&interface=content"
    {"type":"error","status-code":404,"status":"Not Found","result":{"message":"not found"}}

    root@history-episode:~# snap version
    snap    2.37.4-1+b1
    snapd   2.37.4-1+b1
    series  16
    debian  10
    kernel  4.19.0-5-cloud-amd64

vs Ubuntu:

    $ sudo snap version
    [sudo] password for alex:
    snap    2.45.3.1
    snapd   2.45.3.1
    series  16
    ubuntu  20.04
    kernel  5.4.0-45-generic

I think this is related to my (crossed-out) comment here: https://github.com/certbot/certbot/issues/8250#issuecomment-686779748

If you look at https://snapcraft.io/docs/installing-snap-on-debian, you're meant to also do:

    sudo snap install core

And then everything works.

But this is certainly not ideal, because that step is easy to miss.

**Comment by bmw** [comment]
Ah! I'm potentially making noise here unnecessarily then. Sorry about that.

https://snapcraft.io/install/certbot/debian omits that step, but that's a separate (minor) issue if we want to fix it.

**Comment by alexzorin** [comment]
To be clear, I believe the previous implementation (`snap connections certbot` ) has the same issue on Debian 10, because it too was not present in 2.37.

>Ah! I'm potentially making noise here unnecessarily then

I don't think so. Consider that I hit this issue, then you hit it soon after.

Users are easily going to miss this.

But is there anything we can do about it?

**Comment by adferrand** [comment]
I hit it also ... How the Debian instructions are generated in https://snapcraft.io/install/certbot/debian?

**Comment by alexzorin** [comment]
I can think of a half-fix we can do.

In the previous implementation, this plugin detection step failed silently, basically I think because the pipeline is in a subshell that doesn't conclude with a bad exit code:

https://github.com/certbot/certbot/blob/e8518bf206020bc0163f7db33eb7b31126bac6a2/certbot.wrapper#L45

The current pipeline does conclude in a bad exit code, which triggers `set -e`, killing execution entirely:

https://github.com/certbot/certbot/blob/7ddd327f63260c1258e1149a253897d23fa39349/certbot.wrapper#L35

So, either way, plugin detection is broken. 

However, maybe we should allow the script to proceed, despite plugin detection being broken. Eat the bad exit code from `jq` and keep on trucking.

**Comment by alexzorin** [comment]
User hit it on Ubuntu Xenial: https://community.letsencrypt.org/t/issues-running-certbot-command/133347

**Comment by bmw** [comment]
So my main question here is who is affected? At least out of the major OSes, only Debian's tell you to run `sudo snap install core`. Is this because only they need to do it or is it just missing from others? If so, who else? (I'm not looking for a complete list here, I'm just trying to get the scope of this problem.)

> How the Debian instructions are generated in https://snapcraft.io/install/certbot/debian?

It's all controlled by https://github.com/canonical-web-and-design/snapcraft.io which we can submit an issue/PR to.

**Comment by bmw** [comment]
> Maybe we should allow the script to proceed, despite plugin detection being broken. Eat the bad exit code from jq and keep on trucking. We can even print "yo, you need to run sudo snap install core" first.

While I have a motivation to do this to not break existing installations, I personally think I prefer continuing to crash in this case, but improving the error message. Maybe I'll feel differently depending on the scope of the issue, keeping in mind we currently only tell Ubuntu users to use snaps, but I think our snaps are still pretty new. I'd feel differently if our snaps were more widely deployed.

**Comment by alexzorin** [comment]
My understanding of the discussion in Mattermost is that it's not entirely clear how the forum user managed to hit this in Xenial, unless:

- They recently spun up this instance,
- which already had a sufficiently old version of the `core` snap installed,
- and `snapd`'s autorefresh had not had a chance to update the `core` snap yet

In that situation, if they installed the `certbot` snap, then they would end up with this error, at least, until autorefresh kicked in. 

The issue would not occur if any of the following were true:

- The user had upgraded to the available version of `snapd` from `apt` (not 100% on this one), or
- The user did not have a `core` snap installed at all (Certbot would pull the latest one in automatically), or
- The user did `snap refresh` before installing `certbot` (`core` would get updated first), or
- The user waited upto 6 hours for snapd's autorefresh to run (`core` would get updated first)

Based on all of that, it seems to me that a universal documentation-based workaround for this would be to ask the user to do a `snap refresh` before installing.  Either it will do nothing, or it will update any old `core` snap present.

@bmw you mentioned during the call that there was some situation in which installing/updating the `core` snap didn't actually work. I didn't exactly hear all the details due to some network lag, but is that still the case?

**Comment by bmw** [comment]
I'd be interested to know where you found this image. It's older than anything I can quickly find. The oldest AMIs I have all have the core snap preinstalled.

For what it's worth, my opinion is that I suspect this will affect a very small minority of existing users. I think we can fix the problem for future users by fixing our instructions (or make the Certbot snap pull in the core snap if that's possible?).

I also actually prefer the current behavior of erroring out when we're not able to find externally snapped plugins over continuing to run. I think it's better to fail fast and report the problem rather than letting the error propagate later into the program. This change in behavior could break things for existing snap users, but our snaps are so new and the conditions somewhat niche that I really think not many users will be affected. I don't have data on that and if we start seeing a lot of bug reports, I think we should reevaluate.

With this in mind, I would suggest that we do not try to do a release before next week. We're still learning more about what's going on here and we have one seen 1 actual user report a problem here. I personally do not think that this warrants a scrambled Friday release. If people disagree and you need my help to do something tomorrow or over the weekend, people know how to get a hold of me.

**Comment by alexzorin** [comment]
>I'd be interested to know where you found this image. It's older than anything I can quickly find. The oldest AMIs I have all have the core snap preinstalled.

Is the core image being preloaded just an AWS/AMI-specific thing? I don't think any of Canonical's images have any snaps preloaded.

The image in question is the Bionic image from binarylane.com.au. I asked them where they got it, and the owner said it's from https://cloud-images.ubuntu.com, but they haven't updated it in at least a few months (so the image is no longer available). Based on the mtime of `/var/cache/apt/pkgcache.bin`, the image is from March 2019. If either you or Adrian want to have a play on such an unmodified Bionic machine, I added both your SSH keys to `root@state-mister.bnr.la`.

I deployed the latest cloud image on KVM locally and it works fine (pulls new snapd as well). I am still not sure what determines why/when other snaps get pulled. Maybe it is something to ask the snapd maintainers about.

There was [only one other forum report so far](https://community.letsencrypt.org/t/sudo-certbot-certonly-apache-error-jq-error-at-stdin-0-cannot-iterate-over-null-null/133443?u=_az).

**Comment by bmw** [comment]
> Is the core image being preloaded just an AWS/AMI-specific thing?

Nice catch. In addition to the `core` snap, the `amazon-ssm-agent` snap is preinstalled.

Now that it's a new week and we have less of a time crunch, does anyone want to argue for doing an early release to change our snap's behavior here? If we do one, I think the only thing we should do is the minimal fix to (temporarily) revert things to the old behavior of the Certbot snap continuing to run if it cannot check for externally snapped plugins.

I personally still don't think it's worth it. Ubuntu users should only be affected if their packages are pretty old and they have unattended upgrades disabled. Since the release, I have only seen 3 bug reports (counting https://community.letsencrypt.org/t/issues-running-certbot-command/133347/6 as a 3rd) and per the metrics I can see on the snap store, we've had over 4000 new installations. As for users on other OSes, we do not yet direct them towards the snap and the non-Ubuntu OS with the most users is Debian 10 where we only have 192 users.

While I want to make it clear that I do not think we can get exact data on how many users are affected, everything so far makes me think that the scope of the problem is pretty small and while this sudden change in behavior isn't great, I honestly don't think it's the best use of our time to spend at least a couple hours doing an unexpected release. Does anyone disagree?

Once we have this question out of the way, I think we can talk about what changes we want to make outside of the potential minimal bug fix release.

**Comment by adferrand** [comment]
I agree it does not worth an early release.

**Comment by ohemorange** [comment]
Pretty sure this is the relevant bug; maybe this means it's only a problem on 18 and older? https://bugs.launchpad.net/snapd/+bug/1819318

**Comment by bmw** [comment]
Out of the non-EOL'd Ubuntu releases, that seems right. We haven't seen any reports of the problem for Ubuntu 20.04.

**Comment by bmw** [comment]
One question I have whose answer I think will affect our next steps here is the difference between the `core` snap and the `snapd` snap. I haven't researched this as much as I think we should, but it seems to me like the `core` snap has a dual purpose of being the base for people to use in their own snaps that is based on Ubuntu 16.04 and a snap to update the `snapd` internals. I think the `snapd` snap just has the latter purpose. Especially with [Ubuntu 16.04 losing its standard support next year](https://wiki.ubuntu.com/Releases), should we have users install the `core` snap or the `snapd` snap?

Ignoring that question for now, I personally think our next steps here should be both of the following:

1. Help users avoid installing the Certbot snap with an old version of snapd. Probably the best way to do this it update our instructions on https://certbot.eff.org to have people manually update it, but I personally think it's worth a little time to see if there's an easy way for us to automate this (outside of trying to run `snap install snapd` ourselves in our snap which I don't think we should do).
2. Improve the error message for people who somehow get themselves into this situation anyway. I still think it should be fatal but we can certainly make the description of the problem and how to fix it more clear. As part of this, I think we may want to consider rewriting `certbot.wrapper` in Python as suggested by Alex at https://github.com/certbot/certbot/issues/8280#issuecomment-690768231, but I personally think we can leave whether or not this is done up to whoever implements this.

**Comment by alexzorin** [comment]
>Especially with Ubuntu 16.04 losing its standard support next year, should we have users install the core snap or the snapd snap?

If we follow the fix in the Ubuntu bug that @ohemorange found (https://github.com/snapcore/snapd/pull/6778), we should ask users to install+refresh the `snapd` snap. In a way, we would be backporting the fix to pre-2.40, in the form of our own documentation.

Agreed on 1 an 2. The error message needs to contain a standalone explanation and fix, because if the user comes from https://snapcraft.io/install/certbot/debian, they're not going to see our documentation at all.

**Comment by bmw** [comment]
The fact that the snap team chose to install `snapd` is good to know. I wish I knew why [snapcraft's own instructions for Debian](https://snapcraft.io/docs/installing-snap-on-debian) tells users to install the `core` snap rather than the `snapd` snap. Maybe they're just out-of-date? I think this might be worth asking about.

That gives me a 3rd action item that I'll edit into my previous post though:

3. Report a bug against https://github.com/canonical-web-and-design/snapcraft.io asking them to fix the Debian instructions to avoid this problem.

**Comment by ohemorange** [comment]
Just noticed this while I was playing around with some other snap stuff, in the output of `snap list`:

```
Name                      Version              Rev    Tracking         Publisher     Notes
core                      16-2.46.1            9993   latest/stable    canonical✓    core
core18                    20200724             1885   latest/stable    canonical✓    base
core20                    20                   634    latest/stable    canonical✓    base
```

This suggests to me that perhaps "core" is just an overloaded term, and the "core" snap is separate from the "core18/20" used as a base for other snaps.

**Comment by bmw** [comment]
To drop some quick additional notes about my own understanding here in the hope they do more harm than good:

1. See https://snapcraft.io/docs/base-snaps which talks about the different snap bases and the difference between `core` and `core16`.
2. I just tried running `sudo snap install snapd` on Debian 9 and it failed with the output `error: cannot install "snapd": invalid snap type: "snapd"`. `sudo snap install core` worked.
3. As my least helpful note, in the past week I saw a document that I believe was on the snapcraft forum talking about how they were essentially transitioning the `core` snap to be what is currently the `snapd` snap, but I can't quickly find it again. I can help search for it next week if people want.

**Comment by bmw** [comment]
It's old, but here's the proposal I found where they were talking about stripping down the `core` snap: https://forum.snapcraft.io/t/introducing-base-snaps/381

I also updated the original post of this issue with the 3 things I suggested we do in response to this issue since it got buried a bit.

**Comment by alexzorin** [comment]
>I just tried running sudo snap install snapd on Debian 9 and it failed with the output error: cannot install "snapd": invalid snap type: "snapd". sudo snap install core worked.

In response to your last 3 comments and adding some of my own discoveries:

I think we have some problems with Debian if we want to use `snap install snapd`:

1. `snap install snapd` requires snapd to support "base snaps".
1. Stretch (`snapd 2.21`) does not support base snaps at all. I guess that's why the [Debian documentation](https://snapcraft.io/docs/installing-snap-on-debian) alone asks you to to upgrade snapd by installing the `core` snap.
2. Buster (`snapd 2.37`) does support base snaps, but appears to be broken. I can't `snap install snapd` on Buster, it complains that no base snap is installed (which is untrue). There are two similar reports on the snap forum ([1 (`2.37`)](https://forum.snapcraft.io/t/trouble-with-snap-scummvm-on-raspian/19254/5?u=_az), [2 (`2.38`)](https://forum.snapcraft.io/t/cannot-install-snapd-cannot-install-snapd-snap-on-a-model-without-a-base-snap-yet/10945?u=_az)), but both remain unexplained. Looking at snapd git, that error message disappeared in 2.43. I am guessing that some bugs got fixed. Ubuntu is at 2.45+.

@adferrand's PR currently asks the user to install the `core` snap. As far as we know (unless I missed something) this works universally. If we instead ask the user to install the `snapd` snap, we are going to end up with Debian users who didn't follow the Debian installation instructions to the letter and suffer the consequences. 

Unless there is a reason that the `core` snap is bad or is going to stop upgrading snapd in the future, I think we should keep relying on it.

**Comment by bmw** [comment]
> Unless there is a reason that the core snap is bad or is going to stop upgrading snapd in the future, I think we should keep relying on it.

I agree that installing the `core` snap seems to be the best approach for now.

> I also ran across this bit of fun when doing `snap remove certbot` on Buster if snapd wasn't upgraded. We may potentially want to check the version of snapd in the installation (`configure`?) hook, as some things may not get setup quite right if snapd is upgraded *after* `certbot` is installed

That is fun and good to know. My personal opinion is we should leave this out for now and just work on getting our instructions updated to tell people install the `core` snap first, but if people continue to miss that step and/or upgrading `snapd` later causes significant problems, we should see what else we can do here including checking this in a snap hook like you suggested.

**Comment by bmw** [comment]
Just to be safe and get more information, I asked about the plan for the `core` snap on the snapcraft forum at https://forum.snapcraft.io/t/what-is-the-best-way-to-update-snapd/20185.

**Comment by bmw** [comment]
It seems like installing the `core` snap is the way to go.

A `snapd` maintainer suggested a multi-step process that'd save our users some disk space, but `snap install core` should always work now and in the future so I think we should use that for simplicity.

**Comment by alexzorin** [comment]
Regarding action item 3 (fixing https://snapcraft.io/install/certbot/debian), any thoughts on doing the same for ourselves? Something like [this](https://tools.letsdebug.net/certbot-snapd-core.gif).

To be honest, while I think fixing that snapcraft.io page is worthwhile, that page is buried quite deep. I'd be more confident if our own instructions helped our Debian (especially Stretch) users out here.

**Comment by bmw** [comment]
I could have been clearer here, but adding `snap install core` to our own instructions is what I was imagining we'd do in response to the first suggestion in my list of action items that I've moved to the first post of this issue. I suggested doing this unless we can find a good way to pull in the `core` snap automatically when the users install the Certbot snap. I haven't spent any time researching if that's possible and still think it might be worth checking, but I'd be surprised if it is possible.

I was actually thinking we'd add it to our snap instructions unconditionally since we've seen bug reports on systems like Ubuntu. I personally still think we should probably do that, but if we wanted to only do it on Debian which should be the main system that's affected, while the dropdown for Debian users in your gif is nice, I think we can make use of the templating system already in the instruction generator to list this step only on the systems we want.

I had been blocking https://github.com/certbot/certbot/issues/8291 on us resolving this piece of this issue, but I'm going to spend some time today to see if that's really necessary. Perhaps Arch Linux and openSUSE users should have a new enough version of snapd already.
